File: src/leetcode/python/medium/SpiralMatrix_54.py
class SpiralMatrix_54(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix or not matrix[0]:
            return []
        row, col = len(matrix), len(matrix[0])
        left, right = 0, col-1
        top, down = 0, row-1
        ans = []
        while left < right and top < down:
            ans.extend([matrix[top][i] for i in range(left, right)])
            ans.extend([matrix[i][right] for i in range(top, down)])
            ans.extend(matrix[down][i] for i in range(right, left, -1))
            ans.extend(matrix[i][left] for i in range(down, top, -1))
            left, right = left+1, right-1
            top, down = top+1, down-1
        if left == right:
            ans.extend([matrix[i][right] for i in range(top, down+1)])
        elif top == down:
            ans.extend([matrix[down][i] for i in range(left, right+1)])
        return ans

File: src/leetcode/java/SpiralMatrix_54.java
package leetcode.java;

import java.util.LinkedList;
import java.util.List;

public class SpiralMatrix_54 {
  public List<Integer> spiralOrder(int[][] matrix) {
    List<Integer> result = new LinkedList<>();
    if (matrix.length == 0) return result;
    
    int left = 0, right = matrix[0].length - 1;
    int top = 0, down = matrix.length - 1;
    while (left <= right && top <= down) {
      for (int i = left; i <= right; i++) result.add(matrix[top][i]);
      top++;
      for (int i = top; i <= down; i++) result.add(matrix[i][right]);
      right--;
      if (top <= down) {
        for (int i = right; i >= left; i--) result.add(matrix[down][i]);
        down--;
      }
      if (left <= right) {
        for (int i = down; i >= top; i--) result.add(matrix[i][left]);
        left++;
      }
    }
    return result;
  }
}
